controllers/services/auth/auth_controller.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"fmt"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	componentsApi "github.com/opendatahub-io/opendatahub-operator/v2/apis/components/v1alpha1"
+	serviceApi "github.com/opendatahub-io/opendatahub-operator/v2/apis/services/v1alpha1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/cluster/gvk"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions/deploy"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions/render/template"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/handlers"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/predicates"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/reconciler"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/resources"
+)
+
+const (
+	odhDashboardConfigCRDName = "odhdashboardconfigs.opendatahub.io"
+)
+
+// NewServiceReconciler creates a ServiceReconciler for the Auth API.
+func NewServiceReconciler(ctx context.Context, mgr ctrl.Manager) error {
+	_, err := reconciler.ReconcilerFor(mgr, &serviceApi.Auth{}).
+		// operands - owned
+		Owns(&rbacv1.ClusterRoleBinding{}).
+		Owns(&rbacv1.ClusterRole{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		WatchesGVK(gvk.Dashboard).
+		WatchesGVK(
+			gvk.CustomResourceDefinition,
+			reconciler.WithEventHandler(handlers.ToNamed(serviceApi.AuthInstanceName)),
+			reconciler.WithPredicates(predicate.NewPredicateFuncs(func(object client.Object) bool {
+				return object.GetName() == odhDashboardConfigCRDName
+			}))).
```

Comment:
wonder: why we need to watch this CRD.
dashboard can update this CRD, but Auth CR only care about value from adminGroups and allowedGroups.
so the change to other part of the CRD should not trigger reconcile for Auth CR, 
right?

Reply:
This is to trigger the dynamic watching when the `OdhDashbaordConfig` CRD gets deployed as consequence of the dashboard installation, Without this, the controller won't notice the existence of the `OdhDashbaordConfig` resource.

Reply:
what i am thinking are two cases:
- upgrade:  user already has dashboard enabled and has this CRD in the cluster also the CR. Auth read out the value from the CR when Auth controller start up, after that any changes will be done directly in the Auth CR instead of odhDahsboardconfig CR since it is read-only.
- new install: Auth CR does not care about odhDashboardConfig at all, any change for the groups will be done directly in Auth CR.

so in the "upgrade" case, the CRD is in the cluster;    in the "new install" case, regardless dashboard enabled or not later.
and if in the "upgrade" case, dashboard was not enabled (no CRD nor CR) and later gets enabled, should be treated like the "new install" case,
right?

Reply:
It all depends on the other comment, says that is the sync is done in the reconcile loop, then we must watch also the CRD, otherwise the sync will become inconsistent.

Reply:
Also, dashboard will be deprecating the groupsConfig from the dashboardConfig, so after this migration period where we have this sync behavioour this watch will be going away.